README.md
# Prestashop Test Lab

## Motivation

From time to time, I need to upgrade some Prestashop sites and it is, admitedly, a tedious work.

For me it is always faster to test locally and then, once everything works as expected, upload the changes to the hosting provider.

After typing the same commands twice, I decided to automate the process a bit, so I only had to execute one command to get everything ready faster and less prone to human error.

It's far from perfect and it's not 100% automated, still it reduces the need to repeat some commands more than once, so the lazy guy in me is satisfied :-).


## Vagrant

I use Vagrant + libvirt to create the virtual environment for the test lab.

You can, of course, use whatever virtualization technology you like; still, I feel Vagrant is a fast and easy way to get things done.

This is the `Vagranfile` used to deploy the test instance on my computer:

```
Vagrant.configure("2") do |config|
  config.vm.define "DOMAIN"
  config.vm.box = "generic/ubuntu2004"

  config.vm.provider :libvirt do |libvirt|
    libvirt.memory = 8192     # 8 GB RAM
    libvirt.cpus = 4          # 4 vCPUs
  end

  #config.vm.network "private_network", type: "dhcp"
  config.vm.network "private_network", type: "static", ip: "192.168.121.150"

  config.vm.provision "shell", path: "ps_bootstrap.sh", :args => "'DOMAIN'"
end
```

It will spin up a single Ubuntu 20.04 LTS image, with 4 vCPUs and 8 GB RAM - adjust as needed based on your resources.

Keep this in mind if you need to test the very latest version of Prestashop which most likely will require PHP 8.x to run, as it might be easier to spin up a Ubuntu 22.04 LTS image (or newer) than installing PHP 8.x on Ubuntu 20.04 LTS... choose your poison! :-) 

It will also assign the static IP address `192.168.121.150` to the VM.

You can also use DHCP if you prefer. Just uncomment the line for DHCP, and comment the one which defines the static IP, like this:

```
  config.vm.network "private_network", type: "dhcp"
  #config.vm.network "private_network", type: "static", ip: "192.168.121.150"
```

Suggestion: replace `DOMAIN` with something you might find more useful, like your real shop's domain name; e.g.: `my-shop.com` - this will be used to, among other things: generate self-signed SSL certificates and the Apache VirtualHost configuration.

Now you can run either `vagrant up` or `recreate_vm.sh` to get the test lab rolled out.


## ps_bootstrap.sh

This is the script that is called by Vagrant once the VM has been deployed and it does all the heavy lifting for you.

However, before deploying your VM with `vagrant up`, there are some variables that you will need to adjust in order to reflect your environment:

Set the Prestashop version that you want to install for your tests:
```
# PrestaShop version to install
ps_version=1.7.4.2
```

Set the PHP version needed to run the Prestashop version defined above:
```
# PHP version to install
# This blog offers a good overview on version compatibility:
# -> https://www.prestasoo.com/blog/prestashop-php-version
# This link offers a good guidance on how to install PHP:
# -> https://www.techsupportpk.com/2019/12/how-to-install-php-71-72-73-74-on-ubuntu-16-17-18-19.html
ps_php_version=7.4
```

The Test Lab creates 2 databases:
- the production/real database
- a decoy/temporary database

Set here the needed information to connect to the (local) production database:
```
# DB info
#
# Define values for your real DB (the one you will import once Prestashop is installed)
ps_real_db=[PROD_DB]
ps_real_db_user=[PROD_DB_USR]
ps_real_db_pass=[PROD_DB_PASSWD].
ps_real_db_prefix=[PROD_DB_PREFIX]                      # default is usually "ps_"
ps_real_db_sql_dump_file=[PROD_DB_SQL_DUMP_FILE]        # e.g.: myDB_dump.sql 
```
In case your DB prefix still is `ps_`, please define it so in `ps_real_db_prefix`.

These are dummy values for the temporary database. You may drop it once you have imported your data into the `ps_real_db`. 
```
# You don't need to modify these if you don't want to.
# We'll use this DB to get Prestashop installed. 
# It can be dropped later on after importing $ps_real_db
ps_decoy_db=prestashop
ps_decoy_db_user=prestauser
ps_decoy_db_pass="Passw0rd1"
```

The script will generate self-signed SSL-certificates for you. Please, fill out these variables to fit your needs:
```
# SSL certificate data - adjust as needed
ssl_country=XX                      # e.g.: CH
ssl_state=[STATE]                   # e.g.: Zurich
ssl_location=[LOCATION]             # e.g.: Zurich
ssl_org="[MY_ORGANIZATION]"         # e.g.: Umbrella Corp.
ssl_ou="[MY_OU]"                    # e.g.: Project Alice
ssl_cn=$ps_domain
```

I live in Switzerland, so I set my APT mirror to something nearer than the US.
Set yours according to where you live:
```
# Set your closest APT mirror (default: us, set to your country)
ps_apt_mirror=ch
```

Set your time zone; check the link in the comments for further info:
```
# Set your time-zone
# ->  https://www.php.net/manual/en/timezones.php for valid values
ps_timezone="Europe/Zurich"
```


## How to use this thing

At the risk of sounding biased: it's really easy (perhaps because I've done this several times already) and, for sure, it can be improved. Feel free to create a PR with your suggestions.

Once you have defined all the needed variables mentioned above, you can execute either `vagrant up` or use the `./recreate_vm.sh` script. If you use the latter, it will automatically ssh into the VM once it's created.

`ps_bootstrap.sh` will download the Prestashop version you defined in `ps_version` and uncompress it for you under `/var/www/html` and it will also install Apache2.

I recommend you to edit your `/etc/hosts` and add an entry for your domain while you are testing:

`192.168.121.150  my-shop.com                www.my-shop.com      phpmyadmin.my-shop.com`

...so you can connect to it using your web browser as usual. Don't forget to remove/comment it once you're done and want to access the real website.

That IP address is defined in the `Vagrantfile` shown above:

` config.vm.network "private_network", type: "static", ip: "192.168.121.150" `

Once your Vagrant instance is up & running, you will need to upload:
- the latest MySQL/MariaDB dump
- the `img` directory (hint: create a ZIP file: `img.zip`)

Now run the Prestashop install process as usual:
- point your browser to my-shop.com (or whatever domain you defined)
- follow the instructions on screen until you get to the DB configuration screen
- in the DB config screen, enter the info for the dummy DB:

  - db_name: `prestashop`
  - db_user: `prestauser`
  - db_pass: `Passw0rd1`

After that, you can import your MySQL/MariaDB dump onto your `ps_real_db` using the `db_import.sh` script (see below) and update the contents of the `img` directory with the `refresh_img.sh` script (also see below).

Finally, you will need to update the DB connection info in `app/config/parameters.php` to point to the production DB; use the `update_parameters.sh` script for that if you don't want to do it manually.

And you are now all set and should have a local working copy of your prod website.

After you finished with all your tests, you can kill the Vagrant instance with `vagrant destroy`; you may use `-f` if you don't want to be asked for confirmation.

## What if something goes wrong during the deployment?
You can take a look at the log file where everything has been logged under `/home/vagrant/ps_testlab_install_YYYYMMDD-HH:MM:SS.log` to find out where things failed during the installation.



## Helper scripts

The following little scripts were created to make my life a bit easier after the VM was deployed. Perhaps they will help you too... They are very simple, take certain conditions for granted, and don't include any sanity checks (I might improve that though). 

I list them here for your reference:

### recreate_vm.sh

Sometimes you need to start your tests from scratch, this script will destroy your running Vagrant instance and create a fresh one based on the `Vagrantfile` defined above. 

It will NOT ask for confirmation. Use with care.


### create_db_and_user.sh

This script is generated by `ps_bootstrap.sh` and placed under `/root` inside the Vagrant instance.

It is invoked by `ps_bootstrap.sh` to generate all the DB users, DB passwords, and the DB themselves.


### db_import.sh

This script is generated by `ps_bootstrap.sh` and placed under `/root` inside the Vagrant instance.

You can use it to import the MySQL/MariaDB dump file (e.g.: myDB_dump.sql) into your production DB.

You must upload the SQL file from your local PC first to your Vagrant instance.

e.g.: `vagrant upload myDB_dump.sql`

Reminder: the stuff you upload this way lands under `/home/vagrant`

### refresh_img.sh

This script is generated by `ps_bootstrap.sh` and placed under `/root` inside the Vagrant instance.

It will remove the default `img` directory and replace it with the contents of `img.zip` that you must first upload to your Vagrant instance.

e.g.: `vagrant upload img.zip`

Reminder: the stuff you upload this way lands under `/home/vagrant`

### update_parameters.sh

This script is generated by `ps_bootstrap.sh` and placed under `/root` inside the Vagrant instance.

It will take care of updating the DB connection settings in `app/config/parameters.php`.


## bugs

I'm totally aware that my scripts dream of perfection :-). If you find a bug, please raise an issue on GitHub and will take a look at it. 

Ideally, create a PR and submit your own bug fixes! :-) 


## disclaimer
I am providing this test lab kit to the Prestashop Community on an "AS IS" basis.

I will not be held responsible for any damage that its use or misuse may cause.


CHANGELOG

# Changelog

## 20231224
- automatically set PHP version system-wide
- disable default Apache sites

## 20231222
- add ondrej repo for PHP
- add reference link on how to manually install other versions of PHP

## 20231214
- initial release


ps_bootstrap.sh
#!/bin/bash

# ps_bootstrap.sh
# by martinm@rsysadmin.com
#
# Please, see the README.md for a full description.
#

# PrestaShop version to install
ps_version=1.7.4.2

# PHP version to install
# This blog offers a good overview on version compatibility:
# -> https://www.prestasoo.com/blog/prestashop-php-version
# This link offers a good guidance on how to install PHP:
# -> https://www.techsupportpk.com/2019/12/how-to-install-php-71-72-73-74-on-ubuntu-16-17-18-19.html
ps_php_version=7.4

# DB info
#
# Define values for your real DB (the one you will import once Prestashop is installed)
ps_real_db=[PROD_DB]
ps_real_db_user=[PROD_DB_USR]
ps_real_db_pass=[PROD_DB_PASSWD].
ps_real_db_prefix=[PROD_DB_PREFIX]                      # default is usually "ps_"
ps_real_db_sql_dump_file=[PROD_DB_SQL_DUMP_FILE]        # e.g.: myDB_dump.sql 

# You don't need to modify these if you don't want to.
# We'll use this DB to get Prestashop installed. 
# It can be dropped later on after importing $ps_real_db
ps_decoy_db=prestashop
ps_decoy_db_user=prestauser
ps_decoy_db_pass="Passw0rd1"

# Target directory; e.g.: mydomain.tld
# This is used to generate the SSL certs later on
ps_domain=$1
if [ -z $ps_domain ]
then
    echo -e "\nUsage: $(basename $0) <domain-name.tld>"
    echo -e "\n e.g.: $(basename $0) mydomain.tld"
    exit 1
fi

# SSL certificate data - adjust as needed
ssl_country=XX                      # e.g.: CH
ssl_state=[STATE]                   # e.g.: Zurich
ssl_location=[LOCATION]             # e.g.: Zurich
ssl_org="[MY_ORGANIZATION]"         # e.g.: Umbrella Corp.
ssl_ou="[MY_OU]"                    # e.g.: Project Alice
ssl_cn=$ps_domain

# Set your closest APT mirror (default: us, set to your country)
ps_apt_mirror=ch

# Set your time-zone
# ->  https://www.php.net/manual/en/timezones.php for valid values
ps_timezone="Europe/Zurich"        


# ----------------------------------------------------------------------------
# ----------- YOU SHOULD NOT NEED TO EDIT BELOW THIS LINE --------------------
# ----------------------------------------------------------------------------

# output everything to a log file - you will need it if something goes wrong
ps_install_log=./ps_testlab_install_$(date +"%Y%m%d-%T").log
exec > >(tee -i $ps_install_log)
exec 2>&1

ps_url="https://github.com/PrestaShop/PrestaShop/releases/download/${ps_version}/prestashop_${ps_version}.zip"
ps_dir=/var/www/html
ps_vhost_config=/etc/apache2/sites-available/$ps_domain.conf

# main()
cat <<BANNER

Prestashop Test Lab Installer

by martinm@rsysadmin.com
=============================

BANNER

echo -e "\n========= Update apt sources: US->CH ========"
sed -i "s/us./${ps_apt_mirror}./g" /etc/apt/sources.list

echo -e "\n========= Update apt cache ============"
sudo apt-get update

echo -e "\n========= Add APT repo for PHP ============"
sudo apt-get install software-properties-common -y
sudo add-apt-repository ppa:ondrej/php -y
sudo apt update

echo -e "\n=========  Installing Apache 2 and PHP ${ps_php_version} ========= "
sudo apt-get install apache2 php${ps_php_version} \
        libapache2-mod-php${ps_php_version} \
        php${ps_php_version}-gd \
        php${ps_php_version}-mbstring \
        php${ps_php_version}-mysql \
        php${ps_php_version}-curl \
        php${ps_php_version}-intl \
        php${ps_php_version}-zip \
        php-xml \
        php-cli \
        unzip -y

echo -e "\n=========  Set PHP default version to ${ps_php_version} ========="
sudo update-alternatives --set php /usr/bin/php${ps_php_version}

php_config_output=$(php -i | grep -i cli/php.ini)
php_ini_path=$(echo -e "\n$php_config_output" | awk -F\> '{ print $2 }')

echo -e "\n=========  Update PHP options ========= "
echo "-- set file_uploads = On "
sed -i 's/^file_uploads =.*/file_uploads = On/g' $php_ini_path

echo "-- set allow_url_fopen = On"
sed -i 's/^allow_url_fopen =.*/allow_url_fopen = On/g' $php_ini_path

echo "-- set short_open_tag = On"
sed -i 's/^short_open_tag =.*/short_open_tag = On/g' $php_ini_path

echo "-- set memory_limit = 256M"
sed -i 's/^memory_limit =.*/memory_limit = 256M/g' $php_ini_path

echo "-- set cgi.fix_pathinfo = 0"
sed -i 's/^cgi.fix_pathinfo =.*/cgi.fix_pathinfo = 0/g' $php_ini_path

echo "-- set upload_max_filesize = 100M"
sed -i 's/^upload_max_filesize =.*/upload_max_filesize = 100M/g' $php_ini_path

echo "-- set max_execution_time = 360"
sed -i 's/^max_execution_time =.*/max_execution_time = 360/g' $php_ini_path

echo "-- set date.timezone = $ps_timezone"
sed -i 's/^date.timezone =.*/date.timezone = ${ps_timezone}/g' $php_ini_path

echo -e "\n=========  Install MariaDB ========= "
sudo apt-get install mariadb-server -y

echo -e "\n=========  Install Apache2 ========= "
sudo apt-get install apache2 -y

echo -e "\n=========  Stop Apache2  ========= "
sudo systemctl stop apache2

echo -e "\n=========  Upload Apache2 VirtualHost configuration ========= "
sudo cat <<DRAMA > $ps_vhost_config

<VirtualHost *:80>
    ServerAdmin admin@$ps_domain
    ServerName $ps_domain
    ServerAlias www.$ps_domain
    DocumentRoot /var/www/html

    <Directory /var/www/html>
        Options +FollowSymlinks
        AllowOverride All
        Require all granted
    </Directory>

    ErrorLog /var/log/apache2/${ps_domain}-error_log
    CustomLog /var/log/apache2/${ps_domain}-access_log common

    # Redirect all HTTP requests to HTTPS
    RewriteEngine On
    RewriteCond %{HTTPS} !=on
    RewriteRule ^/?(.*) https://%{SERVER_NAME}/\$1 [R=301,L]
</VirtualHost>

<VirtualHost *:443>
    ServerAdmin admin@$ps_domain
    ServerName $ps_domain
    ServerAlias www.$ps_domain
    DocumentRoot /var/www/html

    <Directory /var/www/html>
        Options +FollowSymlinks
        AllowOverride All
        Require all granted
    </Directory>

    ErrorLog /var/log/apache2/prestashop-error_log
    CustomLog /var/log/apache2/prestashop-access_log common

    SSLEngine on

    # SSL certificate and private key files
    SSLCertificateFile /etc/ssl/certs/${ps_domain}.crt
    SSLCertificateKeyFile /etc/ssl/private/${ps_domain}.key

    # Optional: SSLCertificateChainFile for intermediate certificates
    # SSLCertificateChainFile /etc/ssl/certs/intermediate.crt

    # Optional: Enable HSTS for improved security (recommended)
    # Header always set Strict-Transport-Security "max-age=31536000; includeSubDomains; preload"

    # Optional: Disable SSLv3 and TLSv1 for improved security (recommended)
    SSLProtocol all -SSLv2 -SSLv3 -TLSv1 -TLSv1.1

    # Optional: Enable Perfect Forward Secrecy for improved security (recommended)
    # SSLCipherSuite EECDH+AESGCM:EDH+AESGCM

    # Optional: OCSP Stapling for improved security (recommended)
    # SSLUseStapling on
    # SSLStaplingCache "shmcb:/var/run/ocsp(128KB)"

    # Optional: Enable HTTP/2 for improved performance (if supported)
    # Protocols h2 http/1.1

    # Optional: Configure additional security-related settings as needed
</VirtualHost>

DRAMA

echo -e "\n=========  Download Prestashop v$ps_version"
wget -qc $ps_url

echo -e "\n=========  Uncompress Prestashop $ps_version ZIP file ========= "
sudo unzip prestashop_${ps_version}.zip -d $ps_dir

echo -e "\n=========  Remove default index.html ========= "
sudo rm -f $ps_dir/index.html

echo -e "\n=========  Fix ownership of $ps_dir ========= "
sudo chown -R www-data:www-data $ps_dir

echo -e "\n=========  Fix directory permissions of $ps_dir ========= "
sudo find $ps_dir -type d -exec chmod 755 {} \;

echo -e "\n=========  Fix file permissions of $ps_dir ========= "
sudo find $ps_dir -type f -exec chmod 644 {} \;

echo -e "\n=========  Enable Apache2 modules: rewrite, SSL ========= "
sudo a2enmod rewrite
sudo a2enmod ssl

echo -e "\n=========  Enable $ps_domain VirtualHost ========= "
sudo a2ensite $ps_domain

echo -e "\n========= Disable default Apache sites ========="
sudo a2dissite 000-default
sudo a2dissite default-ssl 

echo -e "\n=========  Generate self-signed SSL certificate and key for Apache2 ========= "
sudo openssl req -x509 -nodes -days 365 -newkey rsa:2048 \
    -keyout /etc/ssl/private/${ps_domain}.key \
    -out /etc/ssl/certs/${ps_domain}.crt \
    -subj "/C=$ssl_country/ST=$ssl_state/L=$ssl_location/O=$ssl_org/OU=$ssl_ou/CN=$ps_domain"

echo -e "\n=========  Enable and start Apache2 ========= "
sudo systemctl enable --now apache2

echo -e "\n========= Upload update_parameters.sh ========= "
sudo cat<<PARAM > /root/update_parameters.sh
#!/bin/bash

file=/var/www/html/app/config/parameters.php

echo -e "-- create backup"
cp -v \$file \${file}_org
echo -e "-- update DB name"
sed -i 's/$ps_decoy_db/$ps_real_db/g'        \$file
echo -e "-- update DB user"
sed -i 's/$ps_decoy_db_user/$ps_real_db_user/g'     \$file
echo -e "-- update DB passwd"
sed -i 's/$ps_decoy_db_pass/$ps_real_db_pass/g'        \$file
echo -e "-- update DB table prefix"
sed -i 's/ps_/$ps_real_db_prefix/g'                    \$file

PARAM

echo -e "\n========= Upload DB management script ==========="
cat <<SCRIPT > /root/create_db_and_user.sh
#!/bin/bash
DB=\$1
USER=\$2
PASS=\$3

mysql -uroot <<MYSQL_SCRIPT
CREATE DATABASE \$DB;
CREATE USER '\$USER'@'localhost' IDENTIFIED BY '\$PASS';
GRANT ALL PRIVILEGES ON \$DB.* TO '\$USER'@'localhost';
FLUSH PRIVILEGES;
MYSQL_SCRIPT

echo "Database:   \$DB"
echo "Username:   \$USER"
echo "Password:   \$PASS"

SCRIPT

echo -e "\n========= Upload DB import script ========"
cat <<DBIMPORT > /root/db_import.sh
#!/bin/bash
db_name=$ps_real_db

echo "- import \$db_name"
mysql \$db_name < /home/vagrant/$ps_real_db_sql_dump_file

DBIMPORT

echo -e "\n========= Upload img refresh script ========"
cat <<DBIMPORT > /root/refresh_img.sh
#!/bin/bash

echo "- removing old img directory"
rm -rf /var/www/html/img

echo "- unzipping img.zip to img directory"
unzip -q /home/vagrant/img.zip -d /var/www/html

echo "- setting permissions"
chown -R www-data: /var/www/html/img

DBIMPORT

echo -e "\n========= Set script permissions ========"
scripts="create_db_and_user.sh \
         update_parameters.sh \
         db_import.sh \
         refresh_img.sh"
         
for i in $scripts
do
    echo "--- $i"
    chmod -v 755 /root/$i
done

echo -e "\n========= Create decoy DB [ $ps_decoy_db ] =========="
/root/create_db_and_user.sh $ps_decoy_db $ps_decoy_db_user $ps_decoy_db_pass

echo -e "\n========= Create production DB [ $ps_real_db ] ========="
/root/create_db_and_user.sh $ps_real_db $ps_real_db_user $ps_real_db_pass

echo -e "\n=========  INFO: Upload your img.zip and DB dump to complete the process."
